### DISCUSSION TITLE
USB HID Support for RP2040

### DISCUSSION THREAD

**Discussion by kumaaarrr** [micropython/micropython#11281] [RP2040 / Pico]
Hello, I'm currently working on a project using RP2040 and developing the code in Micropython, I want my RP2040 to be recognized as USB HID device. I have searched a lot and still haven't found any reliable source or library which I could use. Does Micropython officially support HID library? I've seen Circuitpython support HID library for Keyboard and Mouse but nothing related to micropython.
Can anyone help me with this? :>

**Comment by dhalbert** [comment]
See #11077

**Reply by kumaaarrr** [reply]
Hi @dhalbert, I've seen examples for Mouse and Keyboard using Ciruitpython and the one in that thread too. But I want my device to be detected as an Braille Display. So I was looking for more like a generic library which I can customize it.


Also, Let's say I have found a C/C++ library, is it possible to use those for Micropython? Is there any example or tutorial regarding this?

**Reply by dhalbert** [reply]
> Hi @dhalbert, I've seen examples for Mouse and Keyboard using Ciruitpython and the one in that thread too. But I want my device to be detected as an Braille Display. So I was looking for more like a generic library which I can customize it.

CircuitPython allows custom HID report descriptors: https://learn.adafruit.com/custom-hid-devices-in-circuitpython

**Comment by rikkenharp** [comment]
@kumaaarrr 
You can look in the raspberry forum for Noobeepi and Hippy. They made one version for 1.17.
I have it running with version 1.19.1 DIRTY
You can download my firmware from here at you own risc. i have changed some stuff and did not check if i destroyed something.
I know that i destroyd mass storage, so don't try using that :)
Keyboard and Mouse works, have not tested Gamepad and Consumercontrol.
If you want you can download it from here.
[https://dl.orangedox.com/DEMO23](url)
PLEASE READ THE [ READ ME AND CREDITS]
I used this version for my demos.
[https://www.youtube.com/watch?v=gjNTW2CAsCI](url)

**Reply by kumaaarrr** [reply]
Hello @rikkenharp, The thing is I want to show my device as a Braille Display, so I was looking for more like a generic library which I can customize to Braille Display HID. I'm not sure if what I'm asking makes sense, I'm very new to this USB HID thing.

**Reply by rikkenharp** [reply]
Hii...Aaaah! Then i would actually try circuitpython first. It's easier to get the descriptor up and running there, you don't have to build the firmware every time you try something new. You maybe go back to Micropython when you have a running descriptor but need the extra speed. Circuitpython can be slow compared to micropython. Way slower in some cases. Find one of the newer braille descriptors and look into the descriptors of CP and change one to your needs or make a custom. I wish you good luck on your adventure, usb descriptors are no fun.

**Reply by kumaaarrr** [reply]
Hey, so I get what you're saying. I'm actually very new to this USB and HID protocol, is there any way I could get more knowledge on what descriptors I should send or how to customize it. I have just started to understand about HID. is there any example regarding this braille HID implementation?

**Reply by rikkenharp** [reply]
Here is a usb hid braille display descriptor, the descriptor itself starts at page 10. Usb hid is a thing i like to stay away from and only deal with it when i need to. Maybe it helps a bit :)
[https://usb.org/sites/default/files/hutrr78_-_creation_of_a_braille_display_usage_page_0.pdf](url)
